modules/core/src/main/java/org/apache/ignite/internal/ShutdownPolicyHandler.java

```diff
@@ -0,0 +1,484 @@
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.ShutdownPolicy;
+import org.apache.ignite.cluster.ClusterNode;
+import org.apache.ignite.cluster.ClusterState;
+import org.apache.ignite.cluster.ClusterTopologyException;
+import org.apache.ignite.internal.processors.cache.CacheGroupContext;
+import org.apache.ignite.internal.processors.cache.distributed.dht.preloader.GridDhtPartitionFullMap;
+import org.apache.ignite.internal.processors.cache.distributed.dht.preloader.GridDhtPartitionMap;
+import org.apache.ignite.internal.processors.cache.distributed.dht.preloader.GridDhtPreloader;
+import org.apache.ignite.internal.processors.cache.distributed.dht.topology.GridDhtPartitionState;
+import org.apache.ignite.internal.processors.metastorage.DistributedMetaStorage;
+import org.apache.ignite.internal.processors.metastorage.persistence.DistributedMetaStorageImpl;
+import org.apache.ignite.internal.util.IgniteUtils;
+import org.apache.ignite.internal.util.typedef.internal.LT;
+import org.apache.ignite.internal.util.typedef.internal.U;
+
+import static org.apache.ignite.ShutdownPolicy.GRACEFUL;
+import static org.apache.ignite.cache.CacheMode.PARTITIONED;
+import static org.apache.ignite.internal.ShutdownPolicyHandler.GracefulShutdownPolicyHandler.GRACEFUL_CLUSTER_SHUTDOWN_METASTORE_KEY;
+import static org.apache.ignite.internal.ShutdownPolicyHandler.GracefulShutdownPolicyHandler.GRACEFUL_SHUTDOWN_METASTORE_KEY;
+import static org.apache.ignite.internal.SupportFeaturesUtils.IGNITE_DISTRIBUTED_META_STORAGE_FEATURE;
+import static org.apache.ignite.internal.SupportFeaturesUtils.isFeatureEnabled;
+
+/**
+ * Base interface for handling {@link ShutdownPolicy}.
+ */
+public interface ShutdownPolicyHandler {
+    /**
+     * Handle shutdown policy.
+     */
+    void handle();
+
+    /**
+     * Allows to interrupt handling process.
+     */
+    void stopHandling();
+
+    /**
+     * Cleanup meta storage keys that are related to the shutdown policies.
+     *
+     * @param ctx Grid context.
+     */
+    static void cleanupOnActive(GridKernalContext ctx) {
+        if (!isFeatureEnabled(IGNITE_DISTRIBUTED_META_STORAGE_FEATURE))
```

Comment:
This feature was implemented in 2019 or even earlier. I think that all versions that do not support the distributed meta-storage are EOL. However, there are test cases that still test/use this system property, and excluding this check in the pull request leads to weird test failures... Also, there are places in our project that use the distributed meta-storage without checking at all...

I think it would be nice to create a separate Jira ticket to deprecate/remove all that obsolete stuff.